File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:
    std::vector<std::string> generateParenthesis(int n) {
        std::vector<std::string> result;
        generateParenthesisHelper(n, 0, 0, "", result);
        return result;
    }

private:
    void generateParenthesisHelper(int n, int open, int close, std::string current, std::vector<std::string>& result) {
        if (current.length() == 2 * n) {
            result.push_back(current);
            return;
        }

        if (open < n) {
            generateParenthesisHelper(n, open + 1, close, current + '(', result);
        }

        if (close < open) {
            generateParenthesisHelper(n, open, close + 1, current + ')', result);
        }
    }
};

File: 22-generate-parentheses/generate-parentheses.java
import java.util.ArrayList;
import java.util.List;

class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<>();
        generateParenthesisHelper(n, n, "", result);
        return result;
    }

    private void generateParenthesisHelper(int left, int right, String current, List<String> result) {
        if (left == 0 && right == 0) {
            result.add(current);
            return;
        }

        if (left > 0) {
            generateParenthesisHelper(left - 1, right, current + "(", result);
        }

        if (right > left) {
            generateParenthesisHelper(left, right - 1, current + ")", result);
        }
    }
}
